core/src/main/java/co/cask/wrangler/steps/parser/ParseAvro.java

```diff
@@ -0,0 +1,160 @@
+  description = "Parses column as AVRO generic record."
+)
+public class ParseAvro extends AbstractStep {
+  private static final Logger LOG = LoggerFactory.getLogger(ParseAvro.class);
+  private final String column;
+  private final String schemaId;
+  private final String type;
+  private final long version;
+  private Decoder<Record> decoder;
+  private boolean decoderInitialized = false;
+  private SchemaRegistryClient client;
+
+  public ParseAvro(int lineno, String directive, String column, String schemaId, String type, long version) {
+    super(lineno, directive);
+    this.column = column;
+    this.schemaId = schemaId;
+    this.type = type;
+    this.version = version;
+  }
+
+  /**
+   * Executes a wrangle step on single {@link Record} and return an array of wrangled {@link Record}.
+   *
+   * @param records  Input {@link Record} to be wrangled by this step.
+   * @param context {@link PipelineContext} passed to each step.
+   * @return Wrangled {@link Record}.
+   */
+  @Override
+  public List<Record> execute(List<Record> records, final PipelineContext context) throws StepException {
+    List<Record> results = new ArrayList<>();
+
+    if (!decoderInitialized) {
+      // Retryer callable, that allows this step attempt to connect to schema registry service
+      // before giving up.
+      Callable<Decoder<Record>> decoderCallable = new Callable<Decoder<Record>>() {
+        @Override
+        public Decoder<Record> call() throws Exception {
+          client = SchemaRegistryClient.getInstance(context);
+          byte[] bytes;
+          if (version != -1) {
+            bytes = client.getSchema(schemaId, version);
+          } else {
+            bytes = client.getSchema(schemaId);
+          }
+          Schema.Parser parser = new Schema.Parser();
+          Schema schema = parser.parse(Bytes.toString(bytes));
+          if ("json".equalsIgnoreCase(type)) {
+            return new JsonAvroDecoder(schema);
+          } else if ("binary".equalsIgnoreCase(type)) {
+            return new BinaryAvroDecoder(schema);
+          }
+          return null;
+        }
+      };
+
+      // Retryer that retries when there is connection issue or any request / response
+      // issue. It would exponentially back-off till wait time of 10 seconds is reached
+      // for 5 attempts.
+      Retryer<Decoder<Record>> retryer = RetryerBuilder.<Decoder<Record>>newBuilder()
+        .retryIfExceptionOfType(IOException.class)
+        .retryIfExceptionOfType(RestClientException.class)
+        .withWaitStrategy(WaitStrategies.exponentialWait(10, TimeUnit.SECONDS))
+        .withStopStrategy(StopStrategies.stopAfterAttempt(5))
+        .build();
+
+      try {
+        decoder = retryer.call(decoderCallable);
+        if (decoder != null) {
+          decoderInitialized = true;
+        } else {
+          throw new StepException("Unsupported decoder types. Supports only 'json' or 'binary'");
```

Comment:
you already checked for this in TextDirectives class. More reason to put all this validate logic into a common place.